plugins/base/src/main/kotlin/translators/descriptors/SyntheticDescriptorDocumentationProvider.kt

```diff
@@ -0,0 +1,47 @@
+package org.jetbrains.dokka.base.translators.descriptors
+
+import org.jetbrains.dokka.analysis.DokkaResolutionFacade
+import org.jetbrains.dokka.analysis.from
+import org.jetbrains.dokka.base.parsers.MarkdownParser
+import org.jetbrains.dokka.links.DRI
+import org.jetbrains.dokka.model.doc.DocumentationNode
+import org.jetbrains.kotlin.descriptors.DeclarationDescriptor
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.idea.kdoc.resolveKDocLink
+import org.jetbrains.kotlin.resolve.DescriptorFactory
+
+private const val ENUM_VALUEOF_TEMPLATE_PATH = "/dokka/docs/kdoc/EnumValueOf.kt.template"
+private const val ENUM_VALUES_TEMPLATE_PATH = "/dokka/docs/kdoc/EnumValues.kt.template"
+
+ internal class SyntheticDescriptorDocumentationProvider(
+    private val resolutionFacade: DokkaResolutionFacade
+) {
+    fun isDocumented(descriptor: DeclarationDescriptor): Boolean = descriptor is FunctionDescriptor
```

Comment:
Theoretically, I guess we could use DRI, but I'm afraid it would lead to bugs with false positives. Using descriptors is much safer

I also don't see the value in using DRI here - `SyntheticDescriptorDocumentationProvider` is an internal class (so could be refactored any time) and we're trying to generate documentation for descriptors anyway, so why not use it? Maybe you have something specific on your mind for why DRI is more desirable?

Reply:
>  bugs with false positives. 

What bugs do you mean?

> I also don't see the value in using DRI here - SyntheticDescriptorDocumentationProvider is an internal class (so could be refactored any time) and we're trying to generate documentation for descriptors anyway, so why not use it?

It can be more extensible and convenient.  I would like to change `provider` in one place for `PSI`/`Descriptor`.

Also, we have a code duplication in the case of low-level PSI/Descriptors.

Reply:
> What bugs do you mean?

DRI doesn't have any information about inheritance, it doesn't know if the member is synthetic or not, if it's a "special" declaration. If we have DRI only, we can only try to match the name of the methods and their arguments. So any methods that match `values()` and `valueOf(kotlin.String)` would qualify. 

Keep in mind that returned value is typed, so there's no `kotlin.Enum` or anything like that to indicate that return type is an enum - so there's no point looking at it. You could also accept and check that `Documentable` is a `DEnum` (if it's not a top level function), but something as stupid as this can also break it:

```kotlin
enum class CustomEnum {
    FOO;

    @JvmName("blabla")
    fun values(): Array<CustomEnum> {
        return arrayOf()
    }
}
```

or this

```
fun CustomEnum.valueOf(arg: String) {}
```

`DescriptorFactory.isEnumValueOfMethod` checks for all of that and has the advantage that we don't have to maintain this function, it will be adjusted if any changes are introduced.


> I would like to change provider in one place for PSI/Descriptor.

If you change the provider, you'll probably have a different data model, so this class might not even be relevant for you. For instance, the descriptor might already contain documentation for these methods, so there's no need to do it additionally. It all depends, that's why the class is internal and is exclusive to this translator

> Also, we have a code duplication in the case of low-level PSI/Descriptors.

I wouldn't call code duplication something that has nothing in common :) It's not a given that both Kotlin and Java will have these methods indefinitely. If you try to generify logic now and have a common `DRI.isDocumented()` method for both PSI and Descriptors, you'll most likely have to split it in two down the road when `kotlin.Enum.values()` method is Deprecated

Reply:
Sorry, I  have not had a look at the provider properly and skipped a link resolving. It is impossible to resolve links without PSI/Descriptor.



I was confusing a bit of name `*Provider`. I thought you implemented a pattern that is about pluggabillity, especially when there are two classes with similar names. 
Can you redesign it?